### ISSUE TITLE
Change Geometry.equals to have equalsExact semantics

### ISSUE THREAD

**Issue by dr-jts** [locationtech/jts#159]
A long-standing pain point in JTS is that Geometry.equals has the semantics of topological equality, rather than exact equality (which is more usual for Java).  This impacts using JTS in frameworks which rely on equals having standard Java semantics (for example, ORMs such as JPA).

Geometry.equals should be changed to have the semantics of Geomety.equalsExact.  For topological equality the Geometry.equalsTopo method can be used.

This will also ensure that equals is consistent with hashCode, which is expected in Java (but is impossible to provide for topology-based equals).

**Comment by airbreather** [comment]
Wouldn't the more painful part be that [`Geometry.equals(Geometry)`](https://github.com/locationtech/jts/blob/cbc13fd1e5d8d6b17d83107dd4719ca0333683bc/modules/core/src/main/java/org/locationtech/jts/geom/Geometry.java#L1065-L1068) uses the semantics of `equalsTopo`, whereas [`Geometry.equals(object)`](https://github.com/locationtech/jts/blob/cbc13fd1e5d8d6b17d83107dd4719ca0333683bc/modules/core/src/main/java/org/locationtech/jts/geom/Geometry.java#L1131-L1136) uses the semantics of `equalsExact`?

i.e., assuming JTS works the same as NTS, you ought to wind up with this apparent inconsistency, which seems more confusing (to me) than the exact truth value on either line (especially in real-world scenarios where `o2` is only typed as `Object` by virtue of it being a key in a hash-based collection):
```java
static void tst()
{
    Geometry g1 = createNewGeom();
    Geometry g2 = createSimilarGeom(); /* such that g1.equalsTopo(g2) && !g1.equalsExact(g2) */
    Object o2 = g2;
    g1.equals(g2); /* true */
    g1.equals(o2); /* false */
}
```

Also, rather than being impossible*, [`hashCode()`](https://github.com/locationtech/jts/blob/cbc13fd1e5d8d6b17d83107dd4719ca0333683bc/modules/core/src/main/java/org/locationtech/jts/geom/Geometry.java#L1143-L1146) is *already* correct for both topological and "exact" equality: it will always return the same value for two topologically equal instances, therefore it will always return that same value for two "exactly" equal instances.  Admittedly, you could invite a "**better**"** (correct) `hashCode()` implementation (i.e., likely fewer collisions) by restricting the semantics of `Geometry.equals(Object)` to "exact" equality, but just using the extents is sufficient.

**In fact, you can always trivially make a **correct** `hashCode()` implementation by simply doing `return 0;`... all implementations of that method which are more complicated than that are (knowingly or otherwise) voluntarily introducing that extra complexity in exchange for improving performance of hash-based collections that key off of that object, not because it's necessary for correctness.*

***This actually wouldn't be unambiguously "better" overall, though: in many real-world cases, I expect that `getEnvelopeInternal()` will already be cached, so it's blindingly fast today... making it slower by, e.g., scanning the entire coordinate data (which I imagine is the most likely alternative) doesn't seem likely to eliminate enough collisions in "interesting" real-world scenarios to pay for itself.  YMMV.*

**Comment by dr-jts** [comment]
> Also, rather than being impossible*, hashCode() is already correct for both topological and "exact" equality

Yes, correct.  My error.   So this is not a driver for this change.

**Comment by dr-jts** [comment]
> Wouldn't the more painful part be that Geometry.equals(Geometry) uses the semantics of equalsTopo, whereas Geometry.equals(object) uses the semantics of equalsExact?

Yes, exactly.  That's why this is a breaking change.  

The example you give is why this change would improve the API, by removing the un-obvious difference between `equals(Geometry)` and `equals(Object)`.  Or am I missing your point?

**Comment by airbreather** [comment]
> The example you give is why this change would improve the API, by removing the un-obvious difference between `equals(Geometry)` and `equals(Object)`. Or am I missing your point?

I'm missing something in general, I guess.

The way the issue is described, it sounds like the implementation of `Geometry.equals(Geometry)` needs to change (because of the two methods that "`Geometry.equals`" could refer to, this is the only one that hasn't been implemented that way for a while now).

But then as it goes on, it justifies itself as though the method to be changed is `Geometry.equals(Object)` ("frameworks which rely on equals having standard Java semantics" and the note about `hashCode()`).

The description doesn't (and definitely didn't) read like "make the two overloads of the `equals` method behave consistently with one another", which is why I chose to ask for clarification.

**Comment by dr-jts** [comment]
Good point - I will clarify the issue description.

**Comment by AndreasWBartels** [comment]
Please do this change, or implement equalsTopo support for GeometryCollection. 
The current implemtation throws a IllegalArgumentException on instances of GeometryCollection.

**Comment by dr-jts** [comment]
> Please do this change, or implement equalsTopo support for GeometryCollection.
> The current implemtation throws a IllegalArgumentException on instances of GeometryCollection.

The reason there is no `equalsTopo` support (and in fact relate in general) for `GeometryCollection` is that it's too hard to compute the exact topology in certain situations (in particular, lines and polygons all crossing at non-vertices).  So this won't happen, at least until some better algorithm comes along.

Will have to think about the effects of this breaking change before implementing it.

**Comment by aaime** [comment]
Just stumbled into this by doing some PMD related cleanups in GeoTools. 
One of the PMD rules for tests is not to use assertTrue when assertEquals is a logical equivalent, e.g.:
``assertEquals(a,b)``
is better than:
``assertTrue(a.equals(b))``
as ``assertEquals`` has a more descriptive failure message, showing the two values, while assertTrue gives nothing unless a message is explicitly provided.

Now, doing the migration, with a and b being geometries, ended up changing the semantics of the comparison, from equalsTopo to equalsExact, because the assertEquals method in JUnit uses ``Object`` as argument, ends up bidinding the "Geometry.equals(Object)`` implementation. I know the reason why it happens, but still find it confusing.

**Comment by dr-jts** [comment]
@aaime yes, that's the kind of crazy-making thing that happens with the current API semantics.  I take it then that you would be in favour of this change?

**Comment by dr-jts** [comment]
@airbreather Are you in favour of this change?

**Comment by airbreather** [comment]
> @airbreather Are you in favour of this change?

Hm.  I am in favor of removing the inconsistency that I posted earlier, and if we have to change the semantics of one, then `Geometry.equals(Geometry)` seems to be the right one to change, even despite the OGC discrepancy.

I wonder if it might not make more sense to deprecate and start working towards removing `Geometry.equals(Geometry)` entirely?  After all, if it's going to work exactly the same as `Geometry.equals(object)`, then I don't see much of a reason to have a separate method for it other than backwards compatibility.

> When this change is made the existing `Geometry.equalsExact(Geometry)` method will be redundant and can be deprecated.

Personally, I prefer the `equalsExact` spelling, since it makes it clear how the equality check would be carried out.

**Comment by dr-jts** [comment]
> > @airbreather Are you in favour of this change?
> 
> Hm. I am in favor of removing the inconsistency that I posted earlier, and if we have to change the semantics of one, then `Geometry.equals(Geometry)` seems to be the right one to change, even despite the OGC discrepancy.

Good.

> I wonder if it might not make more sense to deprecate and start working towards removing `Geometry.equals(Geometry)` entirely? After all, if it's going to work exactly the same as `Geometry.equals(object)`, then I don't see much of a reason to have a separate method for it other than backwards compatibility.

Agreed. Start with deprecation, then remove.

> > When this change is made the existing `Geometry.equalsExact(Geometry)` method will be redundant and can be deprecated.
> 
> Personally, I prefer the `equalsExact` spelling, since it makes it clear how the equality check would be carried out.

Agreed. Will keep `equalsExact`, since it is clear about what it means.  `equals` will delegate to it.

**Comment by aaime** [comment]
I'm in agreement with any move that resolves the inconsistencies, the current situation is hard explain and reason about. The test that failed after migration from assertTrue to assertEquals really needed the equalsTopo behavior, but we can just update the tests to explicitly use that method when the chanage in JTS happens.

**Comment by dr-jts** [comment]
> The test that failed after migration from assertTrue to assertEquals really needed the equalsTopo behavior, but we can just update the tests to explicitly use that method when the chanage in JTS happens.

Why not update it now, so that it works?

**Comment by aaime** [comment]
Cause I already have spent enough time on this. Those upgrading JTS in GeoTools will have a reason to change the tests (they are not many, I'm not setting a major roadblock in front of them).

**Comment by jnh5y** [comment]
@aaime thanks for pointing that out!  Hopefully @jodygarnett and I will remember when the time comes!
